zrml/swaps/src/lib.rs

```diff
@@ -1208,6 +1238,113 @@ mod pallet {
+                // The mutation should never fail, but if it does, we just assume we
+                // consumed all the weight and rollback the pool.
+                let _ = with_transaction(|| match mutation(pool_id) {
+                    Err(err) => {
+                        log::warn!(
+                            "Arbitrage unexpectedly failed on pool {:?} with error: {:?}",
+                            pool_id,
+                            err,
+                        );
+                        total_weight = total_weight.saturating_add(max_weight_per_pool);
+                        TransactionOutcome::Rollback(err.into())
+                    }
+                    Ok(weight) => {
+                        total_weight = total_weight.saturating_add(weight);
+                        TransactionOutcome::Commit(Ok(()))
+                    }
+                });
+                if index.saturating_add(1) == (pool_count as usize) {
+                    break;
+                }
+            }
+            total_weight
+        }
+
+        // Execute arbitrage on a single pool.
+        pub(crate) fn execute_arbitrage(
+            pool_id: PoolId,
+            max_iterations: usize,
+        ) -> Result<Weight, DispatchError> {
+            let pool = Self::pool_by_id(pool_id)?;
+            let pool_account = Self::pool_account_id(pool_id);
+            let balances = pool
+                .assets
+                .iter()
+                .map(|a| (*a, T::AssetManager::free_balance(*a, &pool_account)))
+                .collect::<BTreeMap<_, _>>();
+            let asset_count = pool.assets.len() as u32;
+            let outcome_tokens = pool.assets.iter().filter(|a| **a != pool.base_asset);
+            let total_spot_price = pool.calc_total_spot_price(&balances)?;
+
+            if total_spot_price > BASE.saturating_add(ARBITRAGE_THRESHOLD) {
+                let (amount, iteration_count) =
+                    pool.calc_arbitrage_amount_mint_sell(&balances, max_iterations)?;
+                T::AssetManager::withdraw(pool.base_asset, &pool_account, amount)?;
+                for t in outcome_tokens {
+                    T::AssetManager::deposit(*t, &pool_account, amount)?;
+                }
+                Self::deposit_event(Event::ArbitrageMintSell(pool_id, amount));
+                Ok(T::WeightInfo::execute_arbitrage_mint_sell(asset_count, iteration_count as u32))
+            } else if total_spot_price < BASE.saturating_sub(ARBITRAGE_THRESHOLD) {
+                let (amount, iteration_count) =
+                    pool.calc_arbitrage_amount_buy_burn(&balances, max_iterations)?;
+                T::AssetManager::deposit(pool.base_asset, &pool_account, amount)?;
+                for t in outcome_tokens {
+                    T::AssetManager::withdraw(*t, &pool_account, amount)?;
+                }
+                Self::deposit_event(Event::ArbitrageBuyBurn(pool_id, amount));
```

Comment:
Same probelm as above, but in reverse: If we destroy outcome tokens, we need to move funds from the prize pool to the liquidity pool, instead of just depositing.

Reply:
423e311131371757d7c7fc303b61a47b639d61d4

Reply:
You check the free balances [here](https://github.com/zeitgeistpm/zeitgeist/blob/428dfe291677260043648d7a3b870f3ebbf027b1/zrml/swaps/src/lib.rs#L1311). So for the pool account as transfer source AFAIK the transfer shouldn't fail. Here you use the market account as source of the transfer. Do we need a free balance check somewhere above too?

Reply:
No, the prize pool must always contain exactly as much ZTG as there are full sets issued. So the transfer should never fail.